Metrics events - add utils to support mixpanel integration and persisting user settings in shell.

janedegtiareva opened the pull request:
<a href="https://gitpod.io/#https://github.com/nearprotocol/near-shell/pull/323"><img src="https://gitpod.io/api/apps/github/pbs/github.com/nearprotocol/near-shell.git/80caf92ab44e4223a978cbf9393122cfd4b8760e.svg" /></a>

mikedotexe commented on utils/eventtracking.js:
```diff
@@ -0,0 +1,83 @@
+const readline = require('readline');
+const settings = require('./settings');
+
+const TRACKING_ENABLED_KEY = 'trackingEnaled';
```
typo here :)

janedegtiareva replied:
eagle eye! fixed.

mikedotexe commented:
In standup I heard that we wanted to reach consensus on something? I don't see any details on what is being asked. Looking at the code, this seems great to me. I think adding events to every command is the only way to go. Can't wait to see this live!
Looks like we've got some linting issues, FYI

mikedotexe commented on commands/create-account.js:
```diff
@@ -31,6 +32,8 @@ module.exports = {
 
 async function createAccount(options) {
+    await eventtracking.track(eventtracking.EVENT_ID_CREATE_ACCOUNT_START, {});
+    console.log("?")
```
?

mikedotexe commented on utils/settings.js:
```diff
@@ -0,0 +1,42 @@
+        if (!fs.existsSync(nearPath)) {
+            fs.mkdirSync(nearPath);
+        }
+        const shellSettingsPath = path.join(nearPath, 'settings');
```
can we have this be `settings.js` please

janedegtiareva replied:
.json would be more appropriate. I don't want to make it .js because it's not actually a javascript file, but I am on board with .json

ilblackdragon commented on test/testsettings:
```diff
@@ -0,0 +1 @@
+{"trackingEnaled":false}
```
I don't think we should include this file to repo

janedegtiareva replied:
How would you propose to handle this in CI if we don't have this file?

janedegtiareva replied:
removed file and started generating from .sh

ilblackdragon commented on utils/settings.js:
```diff
@@ -0,0 +1,42 @@
+// Persistent shell settings
+
+const getShellSettings = () => {
+    const nearPath = path.join(homedir, '.near-config');
```
should this be .near or .near-shell or .neardev?

ilblackdragon replied:
also let's make this a constant? same for settings.js

mikedotexe replied:
> should this be .near or .near-shell or .neardev?

@ilblackdragon I recommended Jane put it here as it fits with the structure I laid out in the NEP.
https://github.com/nearprotocol/NEPs/blob/974887126800be47a58e968fac5b997db7d50f7d/text/0000-near-shell.md#settings-config-and-key-management

```
─ awesome-near-project  ⟵ NEAR dApp
 ├── .near-config       ⟵ Stores shell-experience settings and connection configuration
 │  ├── connections     ⟵ Contains files used to connect and deploy a contract (except keys)
 │  │  ├── default.js   ⟵ Default configuration is for development
 │  │  └── localnet.js  ⟵ Custom connection added by user for localnet
 │  └── settings.js     ⟵ Stores near-shell settings, how shell behaves when run in this project
 └── .near-credentials  ⟵ Previously "neardev" this directory contains private key inforamtion
    └── default         
       └── alice.json   ⟵ NEAR account "alice" has private key information here
```
(the same structure also appears in the home directory)
Based on the [comment](https://github.com/nearprotocol/NEPs/pull/31#issuecomment-606615736) in the shell NEP, I understood that we were not going to put anything in `~/.near` as that's reserved for validator data. We can modify the NEP to be `.neardev` if we want instead of `.near-config`. Currently `neardev` isn't a great name and holds private keys, which the NEP calls to change to a more informative directory `.near-credentials` that folks aren't likely to accidentally revision. Anyway, that's the thought process. Again, we can modify the NEP.

janedegtiareva replied:
moved to consts

ilblackdragon replied:
Do we need configs inside project directory?
I really think we should not store any credentials inside project directory - because that's the easiest way to lose them. E.g. #198 
I also think settings should live in common directory for all projects.

potatodepaulo replied:
@ilblackdragon we don't need credentials, but per-project config is necessary.  At the very least, it's important for making sure that people have updated deps e.g. shell itself. I agree that creds and global config should be in a common dir for everything.

ilblackdragon replied:
@potatodepaulo read the NEP and left comments there - https://github.com/nearprotocol/NEPs/pull/31#pullrequestreview-399439203
High level, I don't think there is need in local configuration folder, especially one that starts with `.` as those directories should be gitignored. `package.json` can contain settings that should be "clonnable".

janedegtiareva replied:
Can we limit the discussion to user defined settings here in order to get the PR in for RL1?

ilblackdragon commented on utils/eventtracking.js:
```diff
@@ -0,0 +1,83 @@
+                const answer = await new Promise((resolve) => {
+               
+                    rl.question(
+                        chalk`We would like to collect data on near shell usage to improve developer experience. {bold.yellow Would you like to opt in (y/n)?}`,
```
we should probably talk more information about what we are collecting and how we are planning to use this.

janedegtiareva replied:
@potatodepaulo could we get some final wording for this?

mikedotexe replied:
How about:
> We would like to collect data on near-shell usage to improve developer experience. We will never send private information. We only collect which commands are run via an anonymous identifier.

potatodepaulo replied:
The more text you put in here, the less people are going to read. For now let’s use Mike’s suggestion, and then we can add a link to a data privacy page with a full explanation later.

mikedotexe commented:
Merge conflict 🎗️

potatodepaulo commented:
@janedegtiareva looks like it's asking CI to opt in and that's why the build is breaking for this PR.

ilblackdragon commented:
@janedegtiareva we for sure need to figure out how to not show this in CI - we will have this problem in all examples.

janedegtiareva commented:
> @janedegtiareva we for sure need to figure out how to not show this in CI - we will have this problem in all examples.

This was temporary not working due to changing some things due to comments. Back to working again.

janedegtiareva commented:
> Merge conflict 🎗️

resolved

janedegtiareva commented:
> @janedegtiareva looks like it's asking CI to opt in and that's why the build is breaking for this PR.

Yeah this was temporarily broken due to commit /addressing comments churn. Back to working now.

mikedotexe commented on utils/eventtracking.js:
```diff
@@ -0,0 +1,85 @@
+    EVENT_ID_TX_STATUS_START: 'shell_tx_status_start',
+    EVENT_ID_TX_STATUS_SUCCESS: 'shell_tx_status_success',
+    EVENT_ID_LOGIN: 'shell_login',
+    EVENT_ID_DEPLOY: 'shell_deploy',
```
not used

janedegtiareva replied:
that's already addressed in the part 2 of this effort which actually does events for all the things. can we let this be to avoid unnecessary churn?

mikedotexe commented on utils/eventtracking.js:
```diff
@@ -0,0 +1,85 @@
+    EVENT_ID_TX_STATUS_SUCCESS: 'shell_tx_status_success',
+    EVENT_ID_LOGIN: 'shell_login',
+    EVENT_ID_DEPLOY: 'shell_deploy',
+    EVENT_ID_DEV_DEPLOY: 'shell_dev_deploy',
```
not used

mikedotexe commented on utils/eventtracking.js:
```diff
@@ -0,0 +1,85 @@
+    EVENT_ID_ACCOUNT_KEYS_SUCCESS: 'shell_account_keys_success',
+    EVENT_ID_TX_STATUS_START: 'shell_tx_status_start',
+    EVENT_ID_TX_STATUS_SUCCESS: 'shell_tx_status_success',
+    EVENT_ID_LOGIN: 'shell_login',
```
not used
